### PR TITLE
Add InitVar 'validate' parameter to Mesh

### PR DESCRIPTION
**Pull request by gatling-nrl**
useful to suppress validation even when loglevel<=DEBUG

### CODE REVIEW

File: skfem/mesh/mesh.py [hunk 1]
```diff
@@ -418,7 +419,8 @@ def __post_init__(self):
             self.t = np.ascontiguousarray(self.t)
 
         # run validation
-        self.is_valid(debug=False)
+        if validate:
+            self.is_valid()
```
Comment (kinnala) [inline skfem/mesh/mesh.py:423]: By omitting debug=False, I think you are causing the time consuming checks to run always. Maybe get rid of debug kwarg and move log level check here?

File: skfem/mesh/mesh.py [hunk 2]
```diff
@@ -34,6 +34,7 @@ class Mesh:
     # However, some algorithms (e.g., adaptive refinement) require switching
     # off this behaviour and, hence, this flag exists.
     sort_t: bool = False
+    validate: InitVar[bool] = True # run validation check if log_level<=DEBUG
```
Comment (kinnala) [inline skfem/mesh/mesh.py:37]: Flake fails here because there should be two spaces before the comment.

File: skfem/mesh/mesh.py (hunk 1 above)
Comment (gatling-nrl) [reply skfem/mesh/mesh.py:424]: I added a logger.DEBUG msg any time is_valid gets called (this should probably stay to alert you that something expensive happened if you forgot debug was on). It seems like is_valid is getting called at least twice per mesh creation.

The traceback is just temporary while I try to figure out where these calls to `is_valid` are coming from.

Comment (gatling-nrl) [reply skfem/mesh/mesh.py:424]: I also added a `raise_` keyworkd to `is_valid`, which is occasionally useful to halt and get the stack trace when troubleshooting

Comment (gatling-nrl) [reply skfem/mesh/mesh.py:424]: Ah, there's a temp mesh in `from_meshio`

https://github.com/kinnala/scikit-fem/blob/926c165a21cacad65855a0c4e15409bf84763af7/skfem/io/meshio.py#L111-L112

Probably only should validate the actual mesh returned.

**Comment by gatling-nrl** [comment]
I don't know what it wants for formatting on line 441... that's not my line. To be honest, that line is _really_ hard to read.

**Comment by gdmcbain** [comment]
> I don't know what it wants for formatting on line 441... that's not my line. To be honest, that line is _really_ hard to read.

What I do in such cases is run `black` on the file, copy the reformatted line, undo the general blackening, replace the offending line with the blackened version.

**Comment by gatling-nrl** [comment]
> What I do in such cases is run black on the file, copy the reformatted line, undo the general blackening, replace the offending line with the blackened version.

I split it into two lines. The extra variable helps communicate intent as well as shorten the line. I've never tried `black`... its looks aggressive :)

**Comment by gdmcbain** [comment]
Yes, it is a bit. If you have worked out the line yourself then that's better; black can be handy when you can't be bothered with that. I had to learn it as it's mandated in other packages; e.g. meshio.

**Comment by gatling-nrl** [comment]
Looks like `InitVar` has changed behavior since `3.7`...

**Comment by gatling-nrl** [comment]
https://bugs.python.org/issue33805

**Comment by gatling-nrl** [comment]
I think that calls to `replace` should have `validate=False` to fix this. Presumably one check is enough, on the original object. Thoughts?

**Comment by kinnala** [comment]
What confuses me is that why is this only visible in the doctests?

**Comment by gatling-nrl** [comment]
Ok, I updated all the calls to `replace` in `mesh.py`.

**Comment by kinnala** [comment]
I think there are plenty of `replace` calls in other places as well. Wasn't the issue in the test coming from `refined`?

File: skfem/mesh/mesh.py [hunk 3]
```diff
@@ -34,6 +34,7 @@ class Mesh:
     # However, some algorithms (e.g., adaptive refinement) require switching
     # off this behaviour and, hence, this flag exists.
     sort_t: bool = False
+    validate: InitVar[bool] = True  # run validation check if log_level<=DEBUG
```
Comment (kinnala) [inline skfem/mesh/mesh.py:37]: How about removing `InitVar` and just using `self.validate` in `__post_init__`?

Comment (gatling-nrl) [reply skfem/mesh/mesh.py:37]: Oh, just let `validate` be a normal property of the object? Sure, that's even simpler!

Comment (gatling-nrl) [reply skfem/mesh/mesh.py:37]: But.... isn't this going to cause excessive validating? I'm pretty sure you want `validate=False` when using `replace`. I don't know all the reasons why to call `replace` on a mesh, but it seems like you'd only want `is_valid` run on whatever the original thing was. What do you want to do?

Comment (kinnala) [reply skfem/mesh/mesh.py:37]: I think it's fine to validate in DEBUG log level just in case. Not all the algorithms we have are perfect.

**Comment by gatling-nrl** [comment]
Ok, took `validate=False` out of the calls to `replace` and made it an normal property of a mesh. So now you'd just use `validate=False` when you knew you were constructing an invalid mesh.

File: skfem/mesh/mesh.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import logging
 
-from dataclasses import dataclass, replace
+from dataclasses import dataclass, replace, InitVar
```
Comment (kinnala) [inline skfem/mesh/mesh.py:3]: Unused import.

Comment (kinnala) [inline skfem/mesh/mesh.py:3]:
```suggestion
from dataclasses import dataclass, replace
```

Comment (gatling-nrl) [reply skfem/mesh/mesh.py]: done

**Pull request merged**